### ISSUE TITLE
Serialization

### ISSUE THREAD

**Issue by KOLANICH** [kaitai-io/kaitai_struct#27]
Hello, you have deserialization. How about serialization?

**Comment by GreyCat** [comment]
We've slowly discussing this issue, but it's obviously not as easy as it looks like. Probably it will be a major feature for something like second major version (v2.0 or something like that), so it'll eventually be there, but don't hold your breath for it.

**Comment by KOLANICH** [comment]
>We've slowly discussing this issue, but it's obviously not as easy as it looks like. 

What are the troubles?

**Comment by GreyCat** [comment]
It is pretty easy for simple fixed (C-style) structures. However, as soon as you start using `instances` that bind certain values to offsets in the stream, it becomes much more complex. A very simple example:

```yaml
seq:
  - id: num_files
    type: u4
  - id: files
    type: file
    repeat: expr
    repeat-expr: num_files
types:
  file:
    seq:
      - id: file_ofs
        type: u4
      - id: file_size
        type: u4
    instances:
      body:
        pos: file_ofs
        size: file_size
```

This .ksy describes very simple file index structure which consists of (num_files) of (file_ofs, file_size) pairs. Each pair describes a "file", which can be accessed by an index using something like:

```
file_contents = archive.files[42].body
```

However, adding a file to such archive is a challenge. Ideally, one would want to do something like:

```
archive.files[42].body = file_contents
```

This should set relevant bound `file_size` automatically to accomodate length of assigned `file_contents` and, what's much more complex, assign `file_ofs` somehow. This is not an easy task: KS has no innate knowledge on how to manage unmapped space in the stream, is it limited or not, should you find some unused spot and reuse it or just expand the stream and effectively append `file_contents` to its end, etc.

What's even harder is that in many cases (i.e. archive files, file systems, etc) you don't want rewrite whole file, but just do some changes (i.e. appending new file to the end of archive, or reusing some pre-reserved padding, or something like that).

Another, may be even simpler example: if you read PNGs, you don't care about checksums. When you write PNGs, you have to generate proper checksums for every block — thus we need block checksumming algorithms and some way to bind it to block.

**Comment by KOLANICH** [comment]
what's the problem to deserialize them, edit, serialize back and then write?

**Comment by GreyCat** [comment]
What exactly do you refer as "them"? File archive example? PNG checksums?

**Comment by KOLANICH** [comment]
>What exactly do you refer as "them"? File archive example? PNG checksums?
Almost anything, including files and checksums. In your example
```repeat-expr: num_files```
and
```id: num_files type: u4"
means that "files has length num_files". When you deserialize, you read ```num_files```, create ```files``` array of length ```num_files``` and read there ```num_files``` structures. When you serealize you need the inverse: update "num_files" with length and then write. What if "num_files" is complex expression, for example the hash? Hashes are meant to be ininvertable, so you are unable to do it. So we shift responsibility to a user and to be able serialize he must to provide both forward and inverse mappings for such expressions, except simple cases when they can be derived automatically.

Now some ideas how to implement this. When processing a KS file it should build a graph of object dependencies, in your case ```num_files <-> files[].length```. Then we apply rules, in this case "scalar <-> array.length" -> "scalar <- array.length" which results in ```num_files <- files[].length``` and check for loops, if they are write error.  When you need to serialize you process the description member by member,  launch lazy evaluation according to the graph and write them.

If you want to minimize count of writing operations you store both versions in memory, create diff, and trynot to touch parts not touched by diff.

**Comment by GreyCat** [comment]
> What if "num_files" is complex expression we cannot derive automatically?

Exactly my thoughts. And actually even that requires us to create some sort of inverse derivation engine. For example, if we have a binding:

```yaml
    - id: body_size
      type: u4
    - id: body
      size: body_size * 4 + 2
```

and we update `body`, we should update `body_size`, assigning `(body.size() - 2) / 4` there. If there are some irreversible bindings (i.e. modulo, hashes, etc) - then, at the very least, we need to detect that situation and allow some extra syntax to make it possible for user to set these inverse dependencies manually.

> We can say that every field has some absolute and actual strengh and that edges can go only in direction of non-increasing `actual strengh` and that `actual strengh` of node is max(max(actual strengh of neighbours), absolute strengh).

I'm sorry, but I fail to understand that. Could you rephrase it, provide some examples, or explain why it is helpful, i.e. which problem we're trying to solve here?

> If you want to minimize count of writing operations you store both versions in memory, create diff, and trynot to touch parts not touched by diff.

The point is that we need to add some extra syntax (or API, or something) to make it possible to do custom space allocation or stuff like that. For example, if you're editing a file system, you can't just always append stuff to the end of it: block device usually has a finite capacity and sooner or later you'll exhaust that and your choice would be to pick and reuse some free blocks in the middle of the block device.

You seem to have good ideas on implementation, would you want to join and help implementing it? Basically anything will help, i.e. .ksy syntax ideas, API ideas, tests for serialization, compiler ideas & code, etc, etc.

**Comment by KOLANICH** [comment]
>And actually even that requires us to create some sort of inverse derivation engine. 

Or take some library. Symolic evaluation is rather studied area of math with lots of paper written. I haven't studied it, so it is possible tha some of the ideas I've mentioned were discussed in them.

>I'm sorry, but I fail to understand that. Could you rephrase it, provide some examples, or explain why it is helpful, i.e. which problem we're trying to solve here?

Let we have an array and a size_t scalar - number of elements in array. What is main in the couple? What defines it entirely? Data in the array do, number is needed only to allow us to read the array correctly. For example in c-strings you don't need number because the convention is to terminate by \0. Add more data to array and you'll have to increase array capacity which means you will have to increase number in order to read it  (and everything after it) correctly. So let array.length have strentgh 1, and scalar 0. Then you have link "files[].length <-> num_files". Let we have another array "filenames[]" and its capacity ```num_files-2```, let we also have ```foo``` and ```bar``` fields with something. Connections : ```filenames[].length <-> num_files```,  ```num_files``` <-> ```foo```,  ```bar``` <-> ```foo``` and ```files[].length``` <-> ```foo```.

Now we start processing. Let show strenghs in brackets, absolute first.
1 files[].length (2,0)
2 files[].length 
files[].length (2,2) itself is processed, go to edges
3 files[].length (2,2) <-> num_files(1,0)
4 files[].length (2,2) <-> num_files(1,1)
2>1 so removing reverse edge and setting to 2
5 files[].length (2,2) -> num_files(1,2)
6 files[].length (2,2) <-> foo(1,0)
7 files[].length (2,2) <-> foo(1,1)
8 files[].length (2,2) -> foo(1,1)
9 files[].length (2,2) -> foo(1,2)
10  num_files(1,2) <-> filenames.length (2,0)
11 num_files(1,2) <-> filenames.length (2,2)
equal, this saves both edges.
12 ```num_files (1,2)``` <-> ```foo(1,2)```
13  foo(1,2) <-> bar(1,0)
14 foo(1,2) <-> bar(1,0)

**Comment by GreyCat** [comment]
> Add more data to array and you'll have to increase array capacity which means you will have to increase number in order to read it (and everything after it) correctly. So let array.length have strentgh 1, and scalar 0.

Ok, then what shall we do in case of the following:

```yaml
seq:
  - id: num_objs
    type: u4
  - id: headers
    type: header
    repeat: expr
    repeat-expr: num_objs
  - id: footers
    type: footer
    repeat: expr
    repeat-expr: num_objs
```

This implies that `headers[]` and `footers[]` shall always have the same number of objects. How are the strengths are assigned in this case and how do we enforce that they have equal number of objects?

**Comment by KOLANICH** [comment]
1 see above example
2 throwing an exception, of course

**Comment by GreyCat** [comment]
I've commited very basic PoC code that demonstrated `seq` serialization in Java. It is available in distinct "serialization" branches. To test it, one'll need:

* [compiler with serialization](https://github.com/kaitai-io/kaitai_struct_compiler/tree/serialization)
* [runtime with serialization](https://github.com/kaitai-io/kaitai_struct_java_runtime/tree/serialization)
* [tests with serialization](https://github.com/kaitai-io/kaitai_struct_tests/tree/serialization)

Obviously, only a few tests were converted, and, to be frank, now only a very basic set of types is supported. Even strings are not implemented right now. Testing is very basic too, one can run `./run-java` and see that two packages are ran: [spec](https://github.com/kaitai-io/kaitai_struct_tests/tree/serialization/spec/java/src/io/kaitai/struct/spec) is "normal", reading tests, and [specwrite](https://github.com/kaitai-io/kaitai_struct_tests/tree/serialization/spec/java/src/io/kaitai/struct/specwrite) are writing tests.

I'd really love to hear any opinions on the API (both runtime & generated), Java implementation (it really pain in the ass, as ByteBuffer does not grow, and you have to preallocate the array, or probably reimplement everything twice with something that grows), etc.

**Comment by GreyCat** [comment]
Strings work, repetitions work, enums work, things are slowly coming to reality :)

**Comment by GreyCat** [comment]
Serialization progresses slowly. Basic in-stream user types and processing on byte types work, various fancy byte type stuff like `terminator` and `pad-right` works, etc, etc.

I've got an idea that might be very simple to implement.

## Step 1: manual checks

Generate `_read`, `_write` and `_check` methods. `_check` runs all the internal format consistency checks to ensure that stuff that will be written will be read back properly. For example:

```yaml
seq:
  - id: len_of_1
    type: u2
  - id: str1
    type: str
    size: len_of_1 * 2 + 1
```

would generate:

```java
    public void _read() {
        this.lenOf1 = this._io.readU2le();
        this.str1 = new String(this._io.readBytes(lenOf1() * 2 + 1), Charset.forName("ASCII"));
    }
    public void _write() {
        this._io.writeU2le(this.lenOf1);
        this._io.writeBytes((this.str1).getBytes(Charset.forName("ASCII")));
    }
    public void _check() {
        if (this.str1.bytes().size() != lenOf1() * 2 + 1)
            throw new FormatConsistencyError("str1 size", this.str1.bytes().size(), lenOf1() + 3);
    }
```

To use this properly, one must manually set both `lenOf1` and `str1`:

```java
r.setStr1("abcde");
r.setLenOf1(2);
r._check(); // should pass, so we're clean to take off
r._write(); // should write consistent data that's guaranteed to be readable back
```

## Step 2: dependent variables

We declare some fields as "dependent", and mark them up in ksy:

```yaml
seq:
  - id: len_of_1
    type: u2
    dependent: (str1.to_b("ASCII").size - 1) / 2
  - id: str1
    type: str
    size: len_of_1 * 2 + 1
```

This means that `len1` becomes a read-only variable, `setLenOf1` setter won't be generated. Instead, it would generate slightly different `_write`:

```java
    public void _write() {
        this.lenOf1 = (str1().getBytes(Charset.forName("ASCII")).size() - 1) / 2;
        this._io.writeU2le(this.lenOf1);
        this._io.writeBytes((this.str1).getBytes(Charset.forName("ASCII")));
    }
```

Obviously, using this boils down to single `r.setStr1("abcde");`.

Any comments / ideas / etc?

**Comment by KOLANICH** [comment]
>Any comments / ideas / etc?

1 move evaluation of expression into a separate method
2 use check and write in generic form without explicit expressions in it
3 Why not to use ```value``` instead of ```dependent```?
4 why do we use ```.to_b("ASCII").size```? String encoding is known, so why not just ```.size```?
5 this ```dependent``` are ugly, it'd be nice to eliminate them, but we need a decision what kind of expressions should be resolved automatically. I guess linear ones will be enough. Another question is how to solve them. There is exp4j, but it'll require some code to build a symbolic solver over it.

**Comment by GreyCat** [comment]
> * move evaluation of expression into a separate method
> * use check and write in generic form without explicit expressions in it, but with a method from 1)

You mean, something like that?

```java
    public void _read() {
        this.lenOf1 = this._io.readU2le();
        this.str1 = new String(this._io.readBytes(_sizeStr1()), Charset.forName("ASCII"));
    }
    public void _write() {
        this._io.writeU2le(this.lenOf1);
        this._io.writeBytes((this.str1).getBytes(Charset.forName("ASCII")));
    }
    public void _check() {
        if (this.str1.bytes().size() != _sizeStr1())
            throw new FormatConsistencyError("str1 size", this.str1.bytes().size(), lenOf1() + 3);
    }
    public int _sizeStr1() {
        return lenOf1() * 2 + 1;
    }
```

Does it bring any benefits? It won't really simplify generation (probably on the contrary), and we'll need to invent tons of names for all that `size`, `if`, `process`, etc expressions.

> * Why not to use `value` instead of `dependent`?

Naming is, of course, still to be discussed. One argument against `value` I have is that `value` is already used for reading in value instances, 

> why do we use .to_b("ASCII").size? String encoding is known, so why not just .size?

Using `size` on a string will give out a length of string in characters. If you'll put some non-ASCII into that string, proper `.to_b("ASCII").size` conversion will give you an exception, while just taking "number of bytes = number of characters" will give you corrupted data.

I guess we could try to do some sort of `.bytesize` method for strings taken verbatim from format definition, where "encoding is known", to save retyping the encoding name. However, it still won't work on modified strings, i.e. it's possible to implement `str1.bytesize` and it won't work with `(str1 + 'x').bytesize` (as the latter is CalcStrType, which lacks any source encoding info by design).

> * this `dependent` are ugly, it'd be nice to eliminate them

First of all, you can't really eliminate them completely in any case. Some functions are just irreversible, and in some cases you'll have more free variables than constraints. For example:

```yaml
seq:
  - id: a
    type: u1
  - id: b
    type: u1
  - id: my_str
    type: str
    size: a + b
```

Even if you have byte size of `my_str`, you can't set both `a` and `b` automatically. Reversing stuff automatically would be more like a syntactic sugar feature, just to save from typing boring stuff where it's possible. In fact, I heavily suspect that we'll cover 95% of cases with very crude logic like `size: a` => `a = attr.size`.

**Comment by KOLANICH** [comment]
>value is already used for reading in value instances,

Yes, that's why I've chosen it. It has the same semantic: deriving a value of a struct member from other ones makes it some kind of an ```instance```, but tied to offset.

>we could try to do some sort of .bytesize

```size``` attribute in a .ksy means size in bytes. So I see no reason it to mean anything else in expression language. 

>First of all, you can't really eliminate them completely in any case.

Yes, we have already discussed this.

>Even if you have byte size of my_str, you can't set both a and b automatically.

Since they are of equal strentgh. This should cause ksc to throw an exception.

>Reversing stuff automatically would be more like a syntactic sugar feature, just to save from typing boring stuff where it's possible.

Not only. One expression can contradict another one, it can cause nasty errors ot can be used as a backdoor. Another side is that we (humans) don't know exact expressions ahead of time. So I propose the following:
1 to have a syntax to provide a manual expression
2 missing exoressions are generated by compiler and inserted into a ksy with another type of expressions
3 human examines ksy output for errors and malicious code
4 verified output is used to generate actual code
5 we would sometimes want to
 * regenerate all non-manual expressions in a ksy
 * check that expressions don't contradict each other

**Comment by ixe013** [comment]
I would like to play with the serialization branch...

Is there a pre-built compiler and Java runtime of the serialization branch available? If not, that's ok, I'll setup a Scala build environment.

**Comment by GreyCat** [comment]
@ixe013 There are no pre-built packages for such experimental branches, but it's relatively straightforward to try it yourself. You don't really need anything beyond [sbt](http://www.scala-sbt.org/) — it would download all the required stuff (including Scala compiler and libraries of the versions needed) automatically. See [dev docs](http://doc.kaitai.io/developers.html#_building_an_universal_zip_package) — it's literally one command to run, like `sbt compilerJVM/universal:packageBin` — et voila.

**Comment by glenn-teillet** [comment]
I was able to build a Read-write version of my ksy,  however the generaed code does not compile as the Reader reads from Int and stores them in Enums as Long (Int->Long) and the _writes() try to write the Enum Long as Int (Long->Int cause error).  

I have these errors: 
The method writeBitsInt(int, long) is undefined for the type KaitaiStream
The method writeU1(int) in the type KaitaiStream is not applicable for the arguments (long)
The method writeU2be(int) in the type KaitaiStream is not applicable for the arguments (long)

**Comment by glenn-teillet** [comment]
I also have errors in every constructor, where it tries to assign _parent received as  a KaitaiStruct in the constructor signature to the _parent without casting to the more specific type.

Type mismatch: cannot convert from KaitaiStruct to KaitaiStruct.ReadWrite

**Comment by GreyCat** [comment]
@glenn-teillet There's still lots of work to do, it's nowhere near production-ready. Before trying to compile read-life ksy files with lots of complicate details, probably we should spend some time getting read-write test infrastructure to work, and, probably, porting that older serialization branch code to modern codebase.

**Comment by KOLANICH** [comment]
Some suggested definitions for serialization spec:

* Internal representation are the objects created by KSC-generated code in program's runtime.
* Finalization is a process of transforming the internal representation that way, that only trivial transformations are left to be done to create a serialized structure.
* An expression is the mapping of a subset of internal representation variables called expression arguments to another variable called expression output.
* A reverse expression is an expressio mapping an original expression's output back to its arguments with respect to the current state of the internal representation (including the arguments).
* Trivial transformations are the ones not involving computing expressions and/or their reverse ones. Examples of trivial transformations are endianess conversions and bit shifts induced by using bit-sized fields.

Comments?

**Comment by FSharpCSharp** [comment]
This is a very interesting subject. What is the current status of this? Will this be pursued in the future? It would be very good if you could use serialization. Especially for binary formats based on bit's this would be a great relief. I hope that the issue will be pursued further and that we can also benefit from it in the future.

**Comment by GreyCat** [comment]
No updates here, latest comments like https://github.com/kaitai-io/kaitai_struct/issues/27#issuecomment-334616207 still reflect current state of affairs. Help is welcome :)

**Comment by cryptocode** [comment]
Definitely a complex issue. For my part, an implementation that generates write methods for each field with a simple type (uint32_t, etc for C++) to a stream would go a long way - I could do the object graph traversal and "finer parts" myself.

**Comment by GreyCat** [comment]
@cryptocode If you're interested in C++ support, then I'd suggest to start with implementing related C++ runtime methods. See [this Java runtime branch](https://github.com/kaitai-io/kaitai_struct_java_runtime/tree/serialization) for generally proposed API.

**Comment by mbrudka** [comment]
@KOLANICH Did you consider a slightly different definition for serialization: 
`psc_{ksy_f}(ssc_{ksy_f}(o)) ≡ o` `∀o ∈ O_f`, `∀ksy_f ∈ KSY_f`, `∀pl ∈ PL`, where `O_f` is a space of structure values in which are valid with respect to format f?

**Comment by mbrudka** [comment]
@KOLANICH Did you consider a relaxed definition for the serialization:
`psc_{ksy_f}(ssc_{ksy_f}(o)) ≡ o ∀o ∈ O_f, ∀ksy_f ∈ KSY_f`, where `O_f` is value space of structures valid with respect to format `f`? This definition enebles KSC to ignore internal structure relations and make KSC user responsible to ensure that `∀o ∈ O_f`.
BTW. Why do you enforce PL to be Turing-complete?

**Comment by mbrudka** [comment]
@KOLANICH Did you consider a relaxed definition for the serialization:
`psc_{ksy_f}(ssc_{ksy_f}(o)) ≡ o ∀o ∈ O_f, ∀ksy_f ∈ KSY_f`, where `O_f` is a space of structure values valid with respect to format `f`? This definition enables `KSC` to ignore internal structure relations and makes `KSC` user responsible to ensure that `∀o ∈ O_f`.
In [construct](https://construct.readthedocs.io/en/latest/index.html]) there is a field [Computed](https://construct.readthedocs.io/en/latest/api/misc.html#construct.Computed) which enables to compute some values during serialization using context. Unfortunatelly, this is Python depended and cannot be directly implemented in any `PL`.
BTW. Why do you enforce `PL` to be Turing-complete?

**Comment by KOLANICH** [comment]
>where `O_f` is a space of structure values valid with respect to format `f`?

I don't quite understand what you mean under this. Do you mean `O_f` be a set of internal representations of a format without self-consistency? Further in this message I assume that you mean this.

There are at least 2 issues with your definition.

1
> `psc_{ksy_f}(ssc_{ksy_f}(o)) ≡ o ∀o ∈ O_f, ∀ksy_f ∈ KSY_f`

This relaxed definition allows a degenerate case where the serializer dumps a serialized into any other format except the original one internal representation and the parser checks if the binary string is a serialized internal representation and deserializes it if it is, otherwise parses. Or smuggles some additional info in the fields without an id. In this case the result will be not belong to `FS_f` and that's why be unreadable by external programs. We surely wanna avoid it. So the binary string should be primary and the internal representation should be secondary.

2 
>This definition enables KSC to ignore internal structure relations and makes KSC user responsible to ensure that `∀o ∈ O_f`.

In fact the goal is to make KSC be responsible for enforcement that the output belongs to `FS_f` and make the serializer derive and fix the internal representation to be self-consistent. This should allow automatic processing without human intervention. We just feed ksys and blobs and get the IRs, we process IRs fully automatically, then we dump IR and get a correct self-cinsistent file which can be processed with other programs.

>In construct there is a field Computed which enables to compute some values during serialization using context. Unfortunatelly, this is Python specific and cannot be directly implemented in any PL.
BTW.

I prefer the expressions for computing to be derived by the compiler. I have used some computer algebra systems (including Open Source ones, but Wolfram Mathematica does far better than anything I saw in Open Source) and saw them being able to find analytic solutions for equations. I guess the expressions we are dealing with in `ksy`s are far simpler and in most cases can be derived automatically. But if they cannot, or the derived ones are inefficient, we need some syntax to explicitly provide them.

>Why do you require `PL` to be Turing-complete?

Because non-turing-complete languages lack loops/recursion which are needed to process structures with `repeat-*`, so most of structs are useless for them (and I guess that KSY may be Turing-complete itself but I have no proof for it). Most of programming languages are Turing-complete too. So just for simplicity.

**Comment by mbrudka** [comment]
@KOLANICH The rest of your reply suggest that you understand what is O_f. I am really glad you see the differences between definition as some may think they are equivalent.  Let me then not define it precisely, but please do it for the sake of clarity and completeness if you find my proposition interesting. If about issues you mentioned:
1.  In general your assumption about ``FS_f`` is still at force, thus the definition does not allow to dump structures into arbitrary strings. Certainly, the definition allows to create frames which are not consistent and may be not recognised by other applications eg. frame with a bad checksum. This is why we assume that users is responsible to ensure that ∀o ∈ O_f.
2. The idea to enforce the semantic consistency on KSC level though appealing is rather an illusion. The consistency is the responsibility of the business logic, but not parsers or builders. It is better to avoid it in KS as it will delay the serialisation feature, then will be hard for users and finally will be continuously fixed to meet new requirements. We just need to accept that some frames will not be consistent..

I am not exactly sure what you mean by ``reverse expressions`` - is this ``reverse polish notation``? But I hope that the serialisation will not depend on the implementation of analytic solution solvers as they are really much bigger piece of software that KS. 

In general you are probably right to require PL to be Turing complete. Nevertheless the assumption is not necessary as long as you do not use it in any formal proof or in the implementation.

Concluding: increasing responsibilities lead many software into a dead-end street. Extensive consistency validation is such path in my opinion. KS may avoid it by sticking directly for transforming bits into structures and (hopefully soon) structures into bits.

**Comment by KOLANICH** [comment]
>I am not exactly sure what you mean by `reverse expression`s

>[A reverse expression is an expression mapping an original expression's output back to its arguments with respect to the current state of the internal representation (including the arguments).](#issuecomment-338590768)

>The consistency is the responsibility of the business logic, but not parsers or builders.

Then we need this feature be disableable.

>It is better to avoid it in KS as it will delay the serialisation feature, then will be hard for users and finally will be continuously fixed to meet new requirements. 

Have you read the thread? Our roadmap to the serialization is to implement and debug manual `reverse expression`s first. When we have it working it would be immediately useful - anyone would be able to start using it, but it may be troublesome: there may be lots of variables in formats and it's very labour-consuming to construct and maintain them for every field and instance. The next step, the most tricky one, is to implement autoderivation of them so we wouldn't have to write and maintain them explicitly in most cases.

>But I hope that the serialisation will not depend on the implementation of analytic solution solvers as they are really much bigger piece of software that KS.

[z3](https://github.com/Z3Prover/z3) is not very big. And I hope that all the heavy lifting can be done in compilation phase.

>Nevertheless the assumption is not necessary as long as you do not use it in any formal proof or in the implementation.

Yes, we have not stated that parsers, serializers and ksy itself exist ;) But if we need them exist for every valid ksy (they don't if information-theoretic crypto is used) we have to assume that the language they are transpiled to is Turing-complete.

**Comment by mbrudka** [comment]
> Have you read the thread

I did not assumed that the thread discussion on ideas is a roadmap. But let it be, so good to hear that you want to restrict yourself to manual _reverse expressions_. 

In my opinion even this feature is not necessary to deploy serialisation. The developer will use the code from KS to built its own application with its own logic  above pure parsers/builders. Practice suggest that this is a better place to implement various validators, computables or adapters. Construct imposed some conventions to express this, but it heavily relied on Python specifics which are not available in all KS target languages. Personally I like ascetic  [protlr](https://www.protlr.com) which just ignores expressions and provide industry ready solution for protocol developers. 

I used some similar features in Construct for years as well as observed how my developers were using it. And it was quite problematic. Not because of the implementation - it was sound, but the architectural decision to do this on parsers level. The order of evaluations, the validity of the context, the scope of visibility for embedded structures led to bugs and excessive debugging which is particularly unpleasant for declarative languages. The result: after some hand-on experience the developers ignored this stuff and implemented computables on the application level eg. checksums.

The implementation of expressions expressible in every KS supported language PL is also not a piece of cake even PL is T-complete :).  Consider for example the abstraction layer for expression representation which will facilitate code generation for Ruby. It is the additional complexity on language KSY bindings.

Take all this as a free advice to avoid unnecessary complexities and implementation.  Use or ignore it. But I hope that your  decisions which will bring builders sooner than later. KS has a potent, but its applications without builders are very limited.

> Yes, we have not stated that parsers, serializers and ksy itself exist ;)

Do they? But wait, I see you have the sketch of a theorem: if PL is T-complete, then for every valid KSY  a parser and a builder may be generated which follow some relations... Time for show the proof to prove the usability of T-completness :).

HTH

**Comment by KOLANICH** [comment]
>But let it be, so good to hear that you want to restrict yourself to manual reverse expressions.

Only as a point to the way to fully-automated serialization.

>Practice suggest that this is a better place to implement various validators, computables or adapters.

It's a lot of programmer's effort. The goal of this project is to automate this effort.

>The order of evaluations, the validity of the context, the scope of visibility for embedded structures led to bugs and excessive debugging which is particularly unpleasant for declarative languages. The result: after some hand-on experience the developers ignored this stuff and implemented computables on the application level eg. checksums.

Could you give more details, such as examples and your thoughts about the causes of troubles.

>KS has a potent, but its applications without builders are very limited.

We I guess we would have "dumb" (without automatic consistency enforcement) builders anyway.

>But wait, I see you have the sketch of a theorem: if PL is T-complete, then for every valid KSY a parser and a builder may be generated which follow some relations...

Quite the contrary: if a PL is not T-complete, a parser and a builder cannot be generated for that language for every valid ksy.

**Comment by duncanfinney** [comment]
+1 for this. Even a hybrid approach would be cool. Maybe it could generate code for the simple C structs and then you have to manually implement it for more complicated things.

**Comment by mbrudka** [comment]
@ckaran Good to read you agree with hybrid approach rather that full automated solution at once.

@KOLANICH Your example somehow illustrates our old discussion. I'd not recommend to generate the code which enforces `a` and `b` to be equal. Your idea to just put `static_assert` (for C++) during serializing of `b` is the right and simple way. Let's trust that user made some efforts to feed Kaitai with good data and generate the failsafe code only when it is so easy as above. This is exactly what I intended when facetiously proposed the [relaxed definition for the serialization](https://github.com/kaitai-io/kaitai_struct/issues/27#issuecomment-385408745).

Having said that, I have to admit sadly that there is still no serialization in Kaitai :(

**Comment by ckaran** [comment]
@KOLANICH I agree with you that optimizing compilers and computer algebra systems are able to find a wider set of self-consistent descriptions, and I believe that having that ability in Kaitai Struct is a worthwhile goal, but I don't think that it is one that can be accomplished within Kaitai Struct exclusively unless Kaitai Struct's language is extended to be fully Turing complete.  Instead, I think that Kaitai Struct should follow the approach that Python Construct has taken, making validation of a large number of use cases easy, while punting  the more difficult problems like you've described back to the end user who will need to write validation code in their language of choice.  As problem areas are discovered, Kaitai Struct can be extended to cover those areas.  This prevents paralysis from waiting for the perfect solution.

**Comment by mbrudka** [comment]
@ckaran Not sure it there is a sense to make declarative ksy a Turing complete language. It is better to follow  Construct approach.
@KOLANICH It's a pity, that my implementation cost is too high - I'd have to learn scala, then understand kaitai internals. Two months of preparation for one month of coding.

**Comment by ckaran** [comment]
@mbrudka I think that there was a misunderstanding; I do not want ksy to be Turing Complete, I just meant that if you want to cover every possible corner case from within Kaitai Struct (KS), then you'll need a Turing-complete language.  I personally prefer the approach taken by python Construct, where the simple/common stuff is handled within Construct, and anything more complicated is handled at a higher level.

@KOLANICH I understand about finding developers for the project!  Like you, I only have time to dream big right now.  That said, I do have a little time to think up ideas, and maybe start planning things out.  Are you willing to work on general plans for what a low-level implementation would look like?  

My initial thought is to follow the serialization methods outlined in the [YAML spec](http://yaml.org/spec/1.2/spec.html#id2762313).  We can use the ideas for the [representation graph](http://yaml.org/spec/1.2/spec.html#id2763754) and the [serialization tree](http://yaml.org/spec/1.2/spec.html#id2765410) pretty much directly.  

The representation graph gives us the lowest-level we need to support; at this level, we assume that the representation graph is correct.  We also only need to support the most basic data types in the union of the recommended [YAML schema](http://yaml.org/spec/1.2/spec.html#Schema) and ksy. 

Transforming this to a serialization tree is where I see the hard-work happening; you'll want the node representing the number of elements in a following sequence to occur before the array does, but the checksum over the entire stream to be calculated and stored at the end.  Doing this in a portable manner implies to me that you need to extend ksy so that you can mark an attribute as having to occur before or after some other attribute, which will allow you to build a serialization tree in the correct order.  This is also where you might have custom code hooks that you call out to; e.g., the value of the checksum node is over some portion of the nodes directly ahead of the checksum node.  

Once you have the serialization tree, the process becomes very mechanical, just walking all the nodes in the tree in the correct order, and writing their contents out.

Thoughts, comments, critiques?

**Comment by S-o-T** [comment]
Any updates on the topic? Is there realization for minimal case viable to generate C++/STL code?

**Comment by GreyCat** [comment]
None that I know of.

**Comment by amrnablus** [comment]
So, I skimmed through the discussion and it seems everyone is making very good points. Undoubtedly, this feature/project is very complex and it needs to be broken down into iterations, here are a few suggestions on how to proceed:

1. We can ignore all sorts of validation in the initial phase. If the client developer needs to validate the serialized implementation they can convert the resulting binary data back to a kaitai struct object and do the validation externally.
2. We can initially delegate some of the work to the client developer. For instance, in this case:
```
meta:
  id: test
  file-extension: bin
seq:
  - id: num_files
    type: u1
  - id: file_entry
    type: u1
    repeat: expr
    repeat-expr: 12

```
The developer can do something like:
```
test.file_entry.append(file_object)
test.num_files += 1

```

And then we can gradually automate some of those functionalities, that would be both easier to implement and can help anyone with simple requirements to use kaitai for serialization.

Feedback is welcome

**Comment by GreyCat** [comment]
@amrnablus Just to be clear: it's not like we lack clarity on *how* to proceed, the question is that it needs to be done.

From a conceptual point, what you suggested here is basically exactly what's been proposed and, to some extent, implemented in that Java PoC branch. Implementing `_write` for attr sequences or implementing write for instances is relatively simple, and basic (manual) validation is actually pretty simple too: it's just a matter of checking certain things, a user will still have to do what you've suggested manually even *with* the validation, i.e.:

```
test.file_entry.append(file_object)
test.num_files += 1
```

The question of implementing this basically boils down to a few things:

1. Updating runtimes (relatively easy)
2. Proposing API changes (easy, but needs to be taken care of for every target language separately)
3. Testing it — we'll need to build a whole new testing infrastructure for writing/validation, which just does not exists now. Using KST (https://github.com/kaitai-io/kaitai_struct/issues/253) will make implementing this easier, but it's still a massive chunk of work (medium-hard).
4. Implementing write seq / write parse instance support (medium).
5. Implementing manual validation (medium).
6. Implementing autocorrection (hard to super hard).

**Comment by KOLANICH** [comment]
BTW, [in my repo there is some code for python runtime that can be repurposed easily for implementing writing numbers](https://github.com/KOLANICH/kaitai_struct_python_runtime/blob/refactoring/kaitaistruct.py#L79).

**Comment by amrnablus** [comment]
Ah thanks, let me take a look at the progress so far and see if I can contribute.

**Comment by dullbananas** [comment]
How much progress has been made on this feature

**Comment by fudgepop01** [comment]
I think there's something incredibly interesting discussed in this talk [[here] (https://youtu.be/AdNJ3fydeao)] that gives me some ideas that (may or may not have) already been brought to the table

It speaks of how svelte (a frontend framework/web-compiler) handles truly reactive data. It sort of takes the javascript `$:` operator (don't ask me what it's supposed to do natively, idk either), and generates something reactive from that. Watch the talk to experience some more that it can do - it's really cool.

I think this is a brilliant way of handling dependencies between variables, and it really helped me visualize what something like that might look like in kaitai struct. Theoretically, could there be a second stage of compilation for the generated files? The first pass is what we have now which generates the parser. 

The change would be placing markers with data about values that relate to each other. It doesn't need to be syntactically correct to the native language - that's for the second pass to handle. For now it just generates the parser along with some type of markers. This means that what's generated at this point is essentially just a long string.

The second pass (perhaps a kaitai_struct_linker or something) would ONLY look for these markers and replace them with the proper code to implement the relationships between them. 

In my (possibly incredibly naivé) mind, this is good because it separates the concern of recompilation and serialization into two different phases that could potentially be toggled. If a parser is already in place and variables just need to be "linked" - just run the kaitai_struct_linker. If there is only a need for a parser, then just run the standard compiler that we already have (with an option toggled to leave out the markers).

...Does this idea make any sense?

**Comment by KOLANICH** [comment]
Current processors architecture in large part is incompatible to serialization. Having 2 functions is enough for very dumb processors, but not enough for smart ones like compression algorithms.

* For compression algorithms we have to introduce a function for getting their arguments from the blob.
* but these arguments arguments are not the only parameters. In fact we can split dictionary from the arguments. This means a yet another method to compute an optimal dictionary shared by muptiple streams.
* And both these methods are specific to compression algos only.

So we need another architecture.

Probably we should eliminate `process` and introduce `types` doing the same instead coupled with typed value-instances #127, templates #135 and chains #196.

It may look like

```yaml
seq: 
  - id: dictionary
    size: 256
  - id: compressed_blob0
    size: 100
  - id: compressed_blob1
    size: 500
instance:
  a_compression:
    type: kaitai.compress.algo
  b_compression:
    type: kaitai.compress.algo
  a:
    io: kaitai.compress.algo.uncompressed._io
    type: a
```

where computation of optimal dictionary should be specified in a separate spec which should be a part of KSC stdlib.

More precisely
```yaml
compression_type:
  params:
    - id: compressed
      type: bytes
      generate: compress(args, dict, uncompressed)
    - id: args
      type: any
      generate:
        - get_args(compressed)
        - compute_optimal_args(uncompressed[i])
    - id: dict
      type: bytes
      generate: compute_optimal_dict(uncompressed[], args) # [] is used to inform the compiler that if all the other args form the key, by which all uncompressed are aggregated using this function. I have thought a bit about Einstein notation, but currently have no idea how much use it has here
    - id: uncompressed
      type: bytes
      generate: decompress(args, dict, compressed)
```

defines an interface of processor type

and

```yaml
kaitai.compress.*: compression_type
```

defines that the user-implemented types under that namespace are compressors

**Comment by jchv** [comment]
FWIW, just to keep conversations up to date, I am working on trying to add support for basic serialization capabilities in various runtimes.

*   C++: kaitai-io/kaitai_struct_cpp_stl_runtime#30
*   Golang: kaitai-io/kaitai_struct_go_runtime#16

It looks largely like the Java runtime at least for the write functions themselves, which is entirely accidental (I had only just looked to see if there were any major differences.) I plan to continue pushing PRs to as many runtimes as possible, please yell at me if I'm doing it wrong. :)

(Sidenote: I am aware for more advanced write functionality it would make sense to work on the compiler and runtime portions in parallel, but for these really basic operations I don't think it matters very much, and it's low hanging fruit.)

**Comment by jchv** [comment]
OK, I think it's time to talk about the interface. I realize there's been some discussion for this already, and a lot of it has focused on reversible expressions. I don't want to get into that; all of that stuff would be very nice but is far into the distance in my opinion. One only needs to look at the amount of progress made on this issue since it was created to realize that there is simply not enough time for going directly into symbolic execution and whatnot from scratch. Today users would benefit from even the most basic serialization support. I think that's what we should aim to implement first, as long as it doesn't impede on the potential future.

**First, setters**. I think any language that has getters or helper methods today should have parallel methods for setting as well, and some helper methods may be useful even in languages that do not currently have getters or helper methods. There's already an issue tracking this, #566. This is probably the *first* thing that should be done, before actually trying to implement writing, since it's not very useful otherwise.

- Add setters for basic primitives. For integers, byte arrays, strings, and almost anything else that compiles down to a language primitive, this should be extremely easy.
- For some kinds of variables, `mutable` variations of getters may be needed. For example, for underlying KaitaiStructs.
- For repeated fields, represented as arrays, it's probably advisable to have helpers for adding new entries as well as getting a mutable underlying value.
- Users of the generated library would be responsible for making sure everything makes sense. KaitaiStruct could generate assertions in the write method that the expressions evaluate to the expected values before writing and throw an exception if not. (This would, for example, catch mistakes where the user forgot to increment a length prefix.) This has been discussed here, but I think it is worth including even in the early stages.

**Secondly, instances**. With non-value instances, the incoming KaitaiStream is saved and reused later to lazily grab objects. This works great for reading, but poses problems for writing. In order to support more advanced use cases, modifications to the kaitai struct schema may be needed. For now, I propose pushing some of the complexity back to the user in exchange for allowing the user to do advanced things today:

- Don't store a KaitaiStream for each KaitaiStruct. Instead, each non-value instance would generate a new Instance class deriving from KaitaiInstance.
- KaitaiInstance would provide a function for writing itself out to a KaitaiStream. The calling generated code would be in charge of seeking.
- The KaitaiInstance class should have a constructor that takes in a KaitaiStream, an offset, and a length. This would lazily store the stream and its offset. Attempting to access the value would cause it to be decoded from the stream. If it is never mutated, the decoding step can be skipped when writing out to stream and a direct copy can be made.
- There should be an instance constructor _that takes the type of the field_. This would lazily store the type itself.
- The KaitaiInstance class should have a static convenience function for loading from memory, just storing it as a memory-backed KaitaiStream. (Static function prevents colliding with constructor if the underlying type happens to be the same as the raw binary type, like `str` in C++.)
- During KaitaiStruct deserialization, KaitaiInstance objects corresponding to each instance are populated with the values read during parsing. Because the values are read at a different stage, this is a potential compatibility break, but can probably be minimized by making it happen as late as possible.
- In some cases it may be preferable for all instances to be fetched into memory prior to writing, for example for writing to a file in-place. A new KaitaiStruct method should be provided that recursively loads all instances and converts them to in-memory instances. For KaitaiStructs without instances, this would be a noop (other than calling it for underlying KaitaiStructs.) They would still be direct copied to the output unless attempting to mutate the underlying value.
- Compatibility can largely be maintained in the interface by providing a separate variable/getter/setter for directly manipulating the instance class versus the data. The setter method for the ordinary type itself can just silently manipulate the instance into a memory instance.
- If the KaitaiInstance is unset during write, an exception is thrown.

Some of this is very tentative since I am still mulling over how this might even work preliminarily, but I think it is time to start discussing it ASAP.

**Thirdly, the IO object.** Once the IO object no longer needs to be stored for each KaitaiStruct, I think that there needs to be the ability to construct an object that is blank. For compatibility, the existing constructor that reads the structural data from a data source should still exist, but it would mostly just serve to call the now-public read method.

I think this needs to be changed, but it may be possible to keep compatibility in most of the APIs anyways. Here's my proposal.

Today:

```cpp
class struct_t : public kaitai::kstruct {
public:
    struct_t(kaitai::kstream* p__io, kaitai::kstruct* p__parent = nullptr, struct_t* p__root = nullptr);

private:
    void _read();

...
};
```

Proposed interface:

```cpp
class struct_t : public kaitai::kstruct {
public:
    // New: A default constructor that sets 'zero values' for everything.
    struct_t();
    // Same signature as before.
    struct_t(kaitai::kistream* p__io, kaitai::kstruct* p__parent = nullptr, struct_t* p__root = nullptr);

    // Now public.
    void _read(kaitai::kistream* p__io, kaitai::kstruct* p__parent = nullptr, struct_t* p__root = nullptr);
    // Write method.
    void _write(kaitai::kostream *p__io, kaitai::kstruct* p__parent = nullptr, struct_t* p__root = nullptr);
...
};
```

The `kistream` would only ever be saved in KaitaiInstances.

Again, this is all preliminary proposals. First thing would be to work on setters. But, I do think it is important to start discussing seriously what the first serialization interface should look like.

**Comment by generalmimon** [comment]
@jchv I'm glad that someone is interested in serialization support! However, your comment suggests that we are at the very beginning and that no progress has been made on this issue. That's **not true** at all. There are Java serialization branches, where most of the basic support has already been implemented.


Let me summarize the implementation progress of this feature in Java (edits from maintainers are welcome):
- `seq` fields
  - [x] integers
    - [x] fixed/switchable/inherited endianess
  - [x] bit-sized integers
  - [x] byte arrays
    - [x] terminator/consume/include
    - [x] check if length does not exceed `size`
  - [x] strings
    - [x] `encoding`
    - [x] `terminator`/`consume`/`include`
    - [x] check if length equals `size`
  - [x] enums
  - [x] repetitions
    - [x] `repeat: expr`
      - [x] check if array length matches `repeat-expr`
    - [x] `repeat: eos`
    - [x] `repeat: until`
      - [x] check if last element satisfies the `repeat-until` condition
  - user types
    - [x] without `size`
    - [x] with `size` (creating a substream)
  - switching
    - [x] basic support
    - type casting
      - case types
        - [x] to user types
        - to primitive types (e.g. int)
          - [x] default case specified (combined type is primitive)
          - [ ] default case missing (combined type is nullable (boxed) - unboxing has to be done before actual casting)
      - [ ] to `byte[]` in default case if `size` is specified and default case is missing
        - [ ] with `process`
- [ ] parse `instances`
- testing
  - [x] basic manual creation and running
  - [ ] generation from KST
  - [ ] automated running and publishing results
- [ ] reverse expressions, autocorrection and other advanced stuff

If you don't understand any of the points, feel free to ask!

Can I ask you, @jhcv, why writing parse instances would require such a complicated solution? I was hoping that reading would stay as it is and writing would be similar to writing `seq` fields, only with some seeking, but I may be naive.

**Comment by jchv** [comment]
As far as I know the only way I have deviated from the path taken so far is I’ve attempted to specify what happens with instances. The rest should be pretty similar to the existing branches. Is there anything particularly egregious there?

There’s two use cases that are important:

1. Editing an existing file. It would be nice if it is possible to say, edit a single instance and write it in-place back to a file, or edit a struct but leave the instances alone. But, it would *also* be rather important to be able to then go and save to a new file entirely, and pull all of the lazy loaded bits as needed to write the new file. Finally, in some cases you just want to parse the full file and close it, so you can, for example, fully rewrite the input file. All of these cases become complicated with instances today, and in the C++ code this relies on holding a kaitai stream.

2. Creating new files from scratch. This use case is completely new, but it’s just as important for me as editing existing structures, if not more. Because of my own personal needs, I’d like to push for this to be supported.

Being able to support any editing would, to me, imply adding setters where there are getters, more flexible instances that can still be lazy loaded but also eagerly loaded if needed, and the ability to create new kaitai structs from scratch.

And small note: I don’t know much about the Java generated code or runtime, so admittedly I am not sure how much this applies to Java. I have focused mostly on C++ as it seems to be one of the more challenging ones to support.

**Comment by kruton** [comment]
Is switching on enum types also broken due to kaitai-io/kaitai_struct_compiler#181? I use [such a construct](https://github.com/kruton/ssh-proto/blob/af82b4015dcf2c471f5c87e71eec06b3637a8c12/src/main/resources/kaitai/ssh.ksy#L39) but it ends up generating code where the field is typed as an `Object` instead of another `KaitaiStruct.ReadWrite` and the calls to `_read()` (and `_write(...)`) fail because of that.

**Comment by generalmimon** [comment]
@kruton:

> Is switching on enum types also broken due to [kaitai-io/kaitai_struct_compiler#181](https://github.com/kaitai-io/kaitai_struct_compiler/pull/181)?

Actually not due to ~~https://github.com/kaitai-io/kaitai_struct_compiler/pull/181~~, but **https://github.com/kaitai-io/kaitai_struct/issues/204**, which has been already fixed, but only on the regular `master` branch, not `serialization`. After releasing 0.9, I plan to merge the `master` branch to `serialization` again, so the fix would be included in `serialization` as well.

**Comment by Holzhaus** [comment]
Any news on this?

**Comment by dullbananas** [comment]
the age of my comment here makes me feel old

**Comment by ShadowDancer** [comment]
I would like to contribute some serialization code for c#, but by reading this issue I see there is not much push for doing so.

From my perspective inability to write files in place, or inability to serialize compression algorithms are niche problems, what 90% of users need (I think), is to write few strings and ints into array of bytes, and then maybe set its length.
Maybe lets start with getting some basic functionality across all runtimes, even without supporting complex stuff like functional expressions, and then move on in iterative fashion?

**Comment by DarkShadow44** [comment]
FWIW, my C implementation of an .ksy compiler at https://github.com/DarkShadow44/UIRibbon-Reversing is capable of (basic) serialization.

Is the branch https://github.com/kaitai-io/kaitai_struct_compiler/tree/serialization the current target for serialization, or where would I need to build on to try and add this?

**Comment by KOLANICH** [comment]
Oh an independent impl of KSC in python! Nice, though I doubt it makes much sense to fragment the ecosystem - it just causes effort wasted for nothing. Scala, in which main KSC is writtrn, is a bit esotheric so I don't want to touch it, so I considered a several times to start an own KSC, but rejected that idea all the times. You know, doing development of a KSC impl seriously means feature parity and compatibility, this requires serious constant effort ... for what, for just be able to extend KSC in python? To do that (and in any other lang that can interface JVM easily) we just need architecture improvements in KS itself, idea of which unfortunately was rejected.  If not seriously involve oneself into a standalone impl ... that means the effort is just wasted.

Anyway, synalysis2kaitai has [some code](https://github.com/kaitaiStructCompile/synalysis2kaitai/blob/master/Synalysis2Kaitai/KSAST.py) that can be useful for you.

**Comment by DarkShadow44** [comment]
Well, I mostly made this implementation because I wanted something usable for C, and quick. Just seemed easier than learning Scala back then, although my goal would still be to implement all features back into KSC so I can retire my implementation. Thanks for you suggestions, but I doubt I want to continue this implementation now that it does what I needed.

I'd rather try to add serialization for Java or C# into KS. There's quite a few technical problems with my current implementation, but for a basic implementation it should be usable as reference.

Main concept of my idea of serialization is a multi-pass approach.
Basically it works like:
1) Dry-run to calculate all the offsets of all data, filling special fields of the struct with offsets
2) Give user option to alter data by filling offsets into all places where it's referenced
3) Actual write run, writing the data into a continuous blob

Part 2) is needed because some formats (like UIRibbon) reference parts of it's contents from multiple places. It's not needed for reading the format, but the proprietary reader of MS needs it - so we need to be able to do that.

Each pass works according to the following formula:
1) Write main seq
2) Write main instances one by one
2a) Write instance seq
2b) Write instance instances
Recurse until everything is written.

That idea if open for discussion, if you want to help review it. This is the only way I see to properly write files.

**Comment by KOLANICH** [comment]
@jemarch

**Comment by sweetgiorni** [comment]
> Let me summarize the implementation progress of this feature in Java (edits from maintainers are welcome):
> 
>     * `seq` fields
>       
>       * [x]  integers
>         
>         * [x]  fixed/switchable/inherited endianess
>       * [x]  bit-sized integers
>       * [x]  byte arrays
>         
>         * [x]  terminator/consume/include
>         * [x]  check if length does not exceed `size`
>       * [x]  strings
>         
>         * [x]  `encoding`
>         * [x]  `terminator`/`consume`/`include`
>         * [x]  check if length equals `size`
>       * [ ]  enums ([kaitai-io/kaitai_struct_compiler#181](https://github.com/kaitai-io/kaitai_struct_compiler/pull/181))
>       * [x]  repetitions
>         
>         * [x]  `repeat: expr`
>           
>           * [x]  check if array length matches `repeat-expr`
>         * [x]  `repeat: eos`
>         * [x]  `repeat: until`
>           
>           * [ ]  check if last element satisfies the `repeat-until` condition ([kaitai-io/kaitai_struct_compiler#183](https://github.com/kaitai-io/kaitai_struct_compiler/pull/183))
>       * user types
>         
>         * [x]  without `size`
>         * [x]  with `size` (creating a substream)
>       * switching
>         
>         * [x]  basic support
>         * type casting
>           
>           * case types
>             
>             * [x]  to user types
>             * to primitive types (e.g. int)
>               
>               * [x]  default case specified (combined type is primitive)
>               * [ ]  default case missing (combined type is nullable (boxed) - unboxing has to be done before actual casting)
>           * [ ]  to `byte[]` in default case if `size` is specified and default case is missing
>             
>             * [ ]  with `process`
> 
>     * [ ]  parse `instances`
> 
>     * testing
>       
>       * [x]  basic manual creation and running
>       * [ ]  generation from KST
>       * [ ]  automated running and publishing results
> 
>     * [ ]  reverse expressions, autocorrection and other advanced stuff
> 

@KOLANICH @GreyCat Perhaps you could use this as a template to create separate tickets for tracking serialization implementation in each language? Maybe group them under a milestone? It would help keep track of progress and reduce the number of people asking for updates in this issue :)

**Comment by mmajor73** [comment]
Can some point me to an example of how to use the new serialization features.

**Comment by Mingun** [comment]
As you can see in the #788 research, the `pad-right` property applied only to bytes from the `terminator` byte up to last byte of the attribute size. What value will be used to fill a gap between end-of-type and a `terminator`?

**Comment by KOLANICH** [comment]
I guess:

1. the previous content that was there, without shifting anything, if something was there
2. the one a programmer explicitly has set. see  #613 for more info. It has been rejected, but it may make sense to reconsider
2. 0 by default

**Comment by Mingun** [comment]
Thinking, actually, that is not a problem, because serialization layer will just not write such values -- it always will write a terminator after type bytes. Situation with a gap between type bytes and a terminator is possible only at parsing and there pad bytes have no meaning

**Comment by KOLANICH** [comment]
I think that serialization should be as much "round-trip" as possible:

1. if we parsed a format into an AST, then serialized back, it should give exactly the same bitstream
2. if we parsed a format into an AST and then modified anything constant-size, the n serialized back, the resulting bitstream should be the same as if we have taken a hex-editor and patched this manually. And in fact the runtime should generate a patch and apply it, so writes are minimized
3. when we deal with non-constant-size stuff everything becomes trickier ...

**Comment by DarkShadow44** [comment]
I like the ideas we discussed before: Normal writing creates a new file from scratch, but after reading a file, add an API to write back a part of the file that wasn't changed. IMHO, we should get a working serializer before we get to more exotic cases like this. In the majority of cases, we don't need to write garbage - it's garbage for a reason.

Am I correct in the assumption that nothing has progressed on the serialization front?

**Comment by DarkShadow44** [comment]
Is there any plans to merge the initial serialization support anytime soon? It's though to base future work on such an old parallel branch.
For my C port, once that's done, I could port over my serialization concept from my old python based compiler, but I'm not sure that would be accepted. Opinions?

**Comment by jchv** [comment]
It’d be really nice to see any motion on this. I think even having pretty broken serialization would be good for a start…

**Comment by jchv** [comment]
The problem is simple: while there has been some work done, I don’t think there is anyone actively working on this. It’s worth understanding that a huge use case for Kaitai is in dissecting data and reverse engineering, where you indeed would not care about serialization. It’d be nice to get things moving, but even if some partial implementation would be acceptable, nobody is working on pushing one forward at the moment. So I guess that would be the first problem to address, if any.

Open source isn’t really a competition, though. If you happen to know of a good alternative that fills this niche, it would be nice to know. On my end, I have a Go library that sort-of overlaps, but aside from being limited to Go, it, too, could use some more maintenance…

**Comment by Mimickal** [comment]
Kaitai's compiler does have a target for Python's Construct library, which does support both serialization and deserialization. I have several open PRs for that compiler target to bring it up to parity with other targets. There are some KSY mechanisms it doesn't quite support yet, but for basic use cases it does work.

**Comment by generalmimon** [comment]
I've been taking notes on various things about the implementation of serialization for over a week now - in the (rather unlikely) event that someone wants to read them, here they are: https://gist.github.com/generalmimon/fc22e97faf1fe4b4edc8279b0caa152d

**Comment by DarkShadow44** [comment]
Are you sure we want things like mixing read/write? What I'd target, at least as first version, is to just have a "write to file" functionality. Might not be perfectly efficient, but a lot easier.

Also, not sure if I missed that in your document, what about structures that are pointed to from different locations? That's why I originally opted for a two pass writing, so you get a change to update the data with real offsets from the dry run before it's actually written.

**Comment by generalmimon** [comment]
@DarkShadow44:

> Are you sure we want things like mixing read/write?

I don't know what you're referring to. Can you be more specific and perhaps quote some parts of the document to ensure that we're on the same page?

> Seems to be a lot easier than what you're going for, just fill the classes and then give the main class the instruction to write itself to file/stream.

What do you think I am going for after reading the document? "Fill the classes and then give the main class the instruction to write itself to file/stream" sounds exactly how I intend the resulting serialization-capable classes to be used. I'm just discussing _what_ happens when you give the main class the instruction and exactly _how_ it's done under the hood.

> what about structures that are pointed to from different locations?

Once again, I have no clue what you mean. This can mean so many things. "Pointed to" is like when you have an positional instance (`instances` + `pos`) that parses a structure located at a particular byte offset in the file? "From different locations" means that the same block of bytes at that byte position in the input binary file is parsed from multiple `instances`, i.e. "locations" in the .ksy specification?

**Comment by DarkShadow44** [comment]
> > Are you sure we want things like mixing read/write?
> 
> I don't know what you're referring to. Can you be more specific and perhaps quote some parts of the document to ensure that we're on the same page?

```
so if you would mix writes and reads within one byte, e.g. writeBitsInt(3) + readBitsInt(5), the readBitsInt will misinterpret what writeBitsInt has just set to the bits and bitsLeft properties, thus it would just do some undefined behavior
```
Guess I missed the later part
```
Besides that, unlike the bXbe - bXle situation which can occur in a real .ksy specification, consecutive mixed {write,read}BitsInt* calls should not happen (once everything is correctly implemented in the compiler). 
```
Ignore that, please

> > Seems to be a lot easier than what you're going for, just fill the classes and then give the main class the instruction to write itself to file/stream.
> 
> What do you think I am going for after reading the document? "Fill the classes and then give the main class the instruction to write itself to file/stream" sounds exactly how I intend the resulting serialization-capable classes to be used. I'm just discussing _what_ happens when you give the main class the instruction and exactly _how_ it's done under the hood.

Seeing
```
        BufferedStruct r = new BufferedStruct() {{
            len1(0x10);
            block1(new BufferedStruct.Block() {{
                number1(0x42);
                number2(0x43);
            }});

            len2(0x8);
            block2(new BufferedStruct.Block() {{
                number1(0x44);
                number2(0x45);
            }});

            finisher(0xee);
        }};
```
it looked like it's written as soon the functions are called.

> > what about structures that are pointed to from different locations?
> 
> Once again, I have no clue what you mean. This can mean so many things. "Pointed to" is like when you have an positional instance (`instances` + `pos`) that parses a structure located at a particular byte offset in the file? "From different locations" means that the same block of bytes at that byte position in the input binary file is parsed from multiple `instances`, i.e. "locations" in the .ksy specification?

Take the following dummy file:
```
meta:
  id: test
  endian: le
seq:
  - id: chunks
    type: chunk
    repeat: expr
    repeat-expr: 4
  - id: pos_chunk_2
    type: u2
types:
  chunk:
    seq:
      - id: len
        type: u1
      - id: data
        size: len
```
And now assume pos_chunk_2 has to be the absolute address (from beginning of file) where the second chunk starts. For reading no problem, but for writing we need to set the value correctly. In this case it's simple, but it can get a lot more complicated.

**Comment by KOLANICH** [comment]
>And now assume `pos_chunk_2` has to be the absolute address (from beginning of file) where the second chunk starts.

`pos_chunk_2` is never used, so it can be arbitrary. But you have detected the right problem: what is the interface at which the system of equations should be stopped being built?

I guess:

* If `pos_chunk_2` was used in this spec, setting it right would have been the responsibility of the code generated from this spec. If this spec was used by some external spec and `pos_chunk_2` was used by that spec as a `pos`, it'd have been a responsibility of that spec to set the chunk as a pos. And sometimes it may be needed to allow a solver to treat imported specs as white boxes. So, I guess, we need a hint about that.

Another issue: we sometimes use `pos`-instances, which byte ranges intersect with other entities. And it is the kinda officially-supported way to implement switcheable endianness sometimes. We need to handle this situation somehow.

**Comment by DarkShadow44** [comment]
> If `pos_chunk_2` was used in this spec, setting it right would have been the responsibility of the code generated from this spec.

Either that, or give the user the ability to update variables with position info before they're written. First calculate all target positions, then let the user update the structure again. Seems easier than implementing a hint system, anyways.

**Comment by generalmimon** [comment]
@DarkShadow44:

> Seeing
> 
> ```java
>         BufferedStruct r = new BufferedStruct() {{
>             len1(0x10);
>             block1(new BufferedStruct.Block() {{
>                 number1(0x42);
>                 number2(0x43);
>             }});
> // ...
> ```
>
> it looked like it's written as soon the functions are called.

Sorry, this was really just an incomplete snippet in a rant about whether or not keep the `set` prefix for setters. That incomplete snippet would just create an empty `BufferedStruct` object and fill it with values. To actually serialize something, you need to call the `_write()` method (and if the `_io` property isn't set to a valid writable stream, the `_write(io)` overload has to be called instead).

If you want to look at a bit more complex use case what the serialization API looks like, see test [`TestExpr2.java:24`](https://github.com/kaitai-io/kaitai_struct_tests/blob/b9cf221c6298d1cac3245462ab38efbabfaf6c55/spec/java/src/io/kaitai/struct/specwrite/TestExpr2.java#L24). It shows an example of reading a file, editing the values (while some changed fields are actually dependencies of value instances, which must be manually invalidated) and writing it to a in-memory stream. The manual calculation of the byte size for the `newIo` stream to write to is also accurate, as explained later.

> And now assume pos_chunk_2 has to be the absolute address (from beginning of file) where the second chunk starts. For reading no problem, but for writing we need to set the value correctly. In this case it's simple, but it can get a lot more complicated. I currently don't see how we can do that using a single writing pass.

My design has a simple answer to all sizes and offsets - calculated and set by the user, checked by KS-generated code (well, if EOF errors count as checking, but I guess they do; larger sizes than needed are always valid). Also, all streams have a fixed size (including the root stream), i.e. there is no "growable" stream.

This is very likely to change in the future to make the serialization process more user-friendly, but I would insist on keeping it for the early version of serialization, because it makes many things simpler (from the perspective of KS compiler, not the user, of course). And yet there are quite a few things that are problematic that you don't even know about, but I have already dealt with many of them.

You'd think that having to set sizes of everything as a user is unmanageable, but it's actually easy - you do it gradually just for the direct fields of the type you're currently processing, and delegate lengths that depend on child types to _their_ respective methods that fill each one of them:

```csharp
    private void Build()
    {
        var cdr = new CoreldrawCdr(new List<KaitaiStream>(), null);
        cdr.RiffChunk = new CoreldrawCdr.RiffChunkType(null, cdr, cdr.M_Root);
        var lenFile = Fill(cdr.RiffChunk);
        ;
    }

    private uint Fill(CoreldrawCdr.RiffChunkType chunk)
    {
        chunk.ChunkId = new[] { (byte)'R', (byte)'I', (byte)'F', (byte)'F' };
        chunk.Body = new CoreldrawCdr.CdrChunkData(null, chunk, chunk.M_Root);
        chunk.LenBody = Fill(chunk.Body);
        chunk.PadByte = GetPadByte(chunk.LenBody);
        return
            (uint)chunk.ChunkId.Length + // chunk_id
            4 + // len_body
            chunk.LenBody + // body
            (uint)chunk.PadByte.Length // pad_byte
        ;
    }

    private uint Fill(CoreldrawCdr.CdrChunkData chunkData)
    {
        // ...
```

You can also use this to calculate offsets, reserve space for `instances` and everything. It's always just adding numbers together and rarely some multiplication, very easy.

So I don't have any concerns about it being particularly limiting. Lots of people in this thread seemed like they would be happy with any simple, stupid solution, and I'm here to provide it.

---

@KOLANICH:

> Another issue: we sometimes use `pos`-instances, which byte ranges intersect with other entities. And it is the kinda officially-supported way to implement switcheable endianness sometimes. We need to handle this situation somehow.

I've already thought of that and resolved it: you can individually disable which positional `instances` you want to write in each object. There is a special setter for it (this line is from `InstanceStd.java` generated from test format [`instance_std.ksy`](https://github.com/kaitai-io/kaitai_struct_tests/blob/b9cf221c6298d1cac3245462ab38efbabfaf6c55/formats/instance_std.ksy)):

```java
    public void setHeader_ToWrite(boolean _v) { _toWriteHeader = _v; }
```

I couldn't come up with a better name for it, but it means "whether **to write** the instance `header` or not". The private `_toWriteHeader` field is `true` by default; if you call `setHeader_ToWrite(false)`, the `header` instance will be ignored (i.e. not written) when you call `_write()`.

I encourage you to explore the `serialization` branches - everything around `instances` is implemented. Even in the context of substreams and the `io` key, which was quite tricky to get it right.

I'm currently working on substreams with `process`, they should be coming soon.

**Comment by DarkShadow44** [comment]
> So I don't have any concerns about it being particularly limiting. Lots of people in this thread seemed like they would be happy with any simple, stupid solution, and I'm here to provide it.

Yeah, me too! Thanks for your work and the explanation, IMHO it will get a lot easier when you have a solid foundation!

**Comment by generalmimon** [comment]
Serialization for Java is done. I started a documentation page for how to use it: https://doc.kaitai.io/serialization.html

**Comment by GreyCat** [comment]
Let me break out a separate discussion with a checklist of what needed to merge serialization into master: https://github.com/kaitai-io/kaitai_struct/issues/1060

**Comment by JocelynLoft** [comment]
Thanks for all the hard work! I'm super interested in getting some implementation for C++, is there an existing effort toward that goal or anything I could help with ?

**Comment by hu-chia** [comment]
I'm a little frustrated using the serialization for Java, everything looked great at first, but when I add the CustomProcessor for encryption and decryption, I got many `BufferOverflowException`s.

The generated code created a lot of `ByteBufferKaitaiStream` instances of length `x_InnerSize`, but the length of the byte array returned by the encryption logic is larger than the original bytes, then the `WriteBackHandler` threw `BufferOverflowException`. I guess I must set the correct `x_InnerSize` before write the `ReadWrite` to `KaitaiStream`, is it possible?

What is the correct way to set `x_InnerSize`?

For now, I'm implementing a `KaitaiStream` which could grow itself when the capacity is insufficient, Is there a better way?

**Comment by hu-chia** [comment]
Another problem, the generated java code has below behavior:

```java
        public void _write_Seq() {
            this._io.writeS4be(lenPayload());
            Compressor _process_payload = newCompressor();
            this._raw_payload = _process_payload.encode(payload());
            // It's basically impossible to predict the length of the data and set lenPayload correctly before performing compression.
            if ((_raw_payload().length != lenPayload()))
                throw new ConsistencyError("payload", _raw_payload().length, lenPayload());
            this._io.writeBytes(_raw_payload());
        }
```

**Comment by generalmimon** [comment]
@hu-chia First of all, thanks for your feedback!

> but the length of the byte array returned by the encryption logic is larger than the original bytes, then the `WriteBackHandler` threw `BufferOverflowException`. I guess I must set the correct `x_InnerSize` before write the `ReadWrite` to `KaitaiStream`

Your observations are correct. As I've already hinted in some of previous comments, the serialization feature deals only with serialization itself so far - it expects a fully populated object tree, including lengths, offsets, CRC-32 checksums and whatnot, which must be given by the user in advance. No auto-deriving of values of auxiliary fields takes place yet [^1]. Also, as you correctly note in your second comment, as soon as you call `_write()`, the writing process is non-interactive - it does not stop at any point to give you the opportunity to supply more input.

To your specific questions:

> What is the correct way to set `x_InnerSize`?

The `*_InnerSize` field is actually the size of the _inner_ substream, which means it is independent of the compression algorithm used. It only depends on the data before performing the compression, which you should be able to determine without problems, just by summing the sizes of the fields you have in the substream.

> ```java
>             this._raw_payload = _process_payload.encode(payload());
>             // It's basically impossible to predict the length of the data and set lenPayload correctly before performing compression.
>             if ((_raw_payload().length != lenPayload()))
>                 throw new ConsistencyError("payload", _raw_payload().length, lenPayload());
>             this._io.writeBytes(_raw_payload());
> ```

Yes, this is a bit tricky. Basically you have to make a test write of the contents that will be compressed, compress it and see how many bytes you get. Let me demonstrate the suggested solution at an example:

```ksy
meta:
  id: zlib_serialization
  endian: le
seq:
  - id: len_body
    type: u4
  - id: body
    size: len_body
    type: contents
    process: zlib
types:
  contents:
    seq:
      - id: foo
        type: s4
      - id: bar
        type: f4
```

One feasible way to serialize this is as follows (using the same pattern as in https://github.com/kaitai-io/kaitai_struct/issues/27#issuecomment-1358689992):

```java
    private static byte[] build() throws IOException {
        ZlibSerialization r = new ZlibSerialization();
        r.body = new ZlibSerialization.Contents(null, r, r._root());
        r.setBody_InnerSize(fill(r.body));

        // test write
        byte[] bodyBuf = new byte[r.body_InnerSize()];
        try (KaitaiStream io = new ByteBufferKaitaiStream(bodyBuf)) {
            r.body._write(io);
        }
        r.lenBody = KaitaiStream.unprocessZlib(bodyBuf).length;
        r._check();

        int lenFile =
            4 + // len_body
            (int)r.lenBody // body
        ;
        byte[] output = new byte[lenFile];
        try (KaitaiStream io = new ByteBufferKaitaiStream(output)) {
            r._write(io);
        }
        return output;
    }
    
    private static int fill(ZlibSerialization.Contents cont) {
        cont.setFoo(-4);
        cont.setBar(1.5f);
        cont._check();
        return
            4 + // foo
            4 // bar
        ;
    }

    private static String byteArrayToHex(byte[] arr) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < arr.length; i++) {
            if (i > 0)
                sb.append(' ');
            sb.append(String.format("%02x", arr[i]));
        }
        return sb.toString();
    }

    public static void main(String[] args) throws IOException {
        System.out.println(byteArrayToHex(build()));
        // output: 10 00 00 00 78 9c fb f3 ff ff 7f 06 86 03 f6 00 1b 95 04 f9
    }
```

For CRC-32 checksums, for example, we can use the same approach, just instead of checking how many compressed bytes we get, we calculate the CRC-32 checksum and save it in the object tree so that it's available once it comes to the actual serialization.

Of course, this requires the processing routine to be deterministic.

[^1]: the only exception are [stream parameters](https://doc.kaitai.io/serialization.html#_stream_parameters), because there is indeed no way that the user could set a stream parameter to the "correct" stream, as (sub)streams are created only during the uninterruptible `_write()` execution.

**Comment by hu-chia** [comment]
I guess there would be two compressions here, Am I right?

```java
        r.lenBody = KaitaiStream.unprocessZlib(bodyBuf).length;
        r._check();
```

**Comment by generalmimon** [comment]
To people who are subscribed to this issue: in case you missed it, Kaitai Struct 0.11 was recently released, featuring serialization support in Java and Python (see https://kaitai.io/news/2025/09/07/kaitai-struct-v0.11-released.html). For information on how to use it, see the [Serialization guide](https://doc.kaitai.io/serialization.html).
